File: src/python/laboneq/compiler/seqc/shfppc_sweeper_config_tracker.py
# Copyright 2024 Zurich Instruments AG
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

import attr

from laboneq.compiler.common.shfppc_sweeper_config import (
    SweepCommand,
    SHFPPCSweeperConfig,
)


@attr.define(kw_only=True, slots=True)
class SweepStackFrame:
    count: int
    items: list[SweepCommand | SweepStackFrame] = attr.field(factory=list)


class SHFPPCSweeperConfigTracker:
    def __init__(self):
        self._command_tree: SweepStackFrame = SweepStackFrame(count=1)
        self._stack: list[SweepStackFrame] = [self._command_tree]
        self._ppc_device: str | None = None
        self._ppc_channel: int | None = None

    def has_sweep_commands(self) -> bool:
        return self._ppc_device is not None and self._ppc_channel is not None

    @property
    def _current_frame(self) -> SweepStackFrame:
        return self._stack[-1]

    def add_step(self, ppc_device: str, ppc_channel: int, **kwargs):
        self._current_frame.items.append(SweepCommand(**kwargs))
        self._ppc_device = ppc_device
        self._ppc_channel = ppc_channel

    def enter_loop(self, count: int):
        new_frame = SweepStackFrame(count=count)
        self._current_frame.items.append(new_frame)
        self._stack.append(new_frame)

    def exit_loop(self):
        closed_frame = self._stack.pop()
        # if the frame we are closing is empty, we don't need to keep it
        if len(closed_frame.items) == 0:
            self._current_frame.items.pop()

    def ppc_channel(self) -> int:
        assert self._ppc_channel is not None, "must add a step first"
        return self._ppc_channel

    def ppc_device(self) -> str:
        assert self._ppc_device is not None, "must add a step first"
        return self._ppc_device

    def _flatten(self, command_tree: SweepStackFrame) -> SHFPPCSweeperConfig:
        """Remove any nesting from the command tree

        After flattening, `items` will only hold `SweepCommand`s. The repetition
        count may get adjusted to efficiently represent any internal loops.
        """

        commands: list[SweepCommand | SHFPPCSweeperConfig] = []
        count = command_tree.count

        for c in command_tree.items:
            if isinstance(c, SweepStackFrame):
                assert len(c.items) > 0, "empty frames should have been dropped earlier"
                commands.append(self._flatten(c))
            else:
                commands.append(c)

        if len(commands) == 1 and isinstance(commands[0], SHFPPCSweeperConfig):
            child = commands[0]
            commands = child.commands
            count *= child.count
        else:
            if any(isinstance(c, SHFPPCSweeperConfig) for c in commands):
                raise AssertionError("Nesting rolled loops is not supported")
                # If we wanted to support them, we have to unroll them here.

        assert all(isinstance(c, dict) for c in commands), "items must be SweepCommand"
        return SHFPPCSweeperConfig(
            count=count,
            commands=commands,
            ppc_device=self.ppc_device(),
            ppc_channel=self.ppc_channel(),
        )

    def finish(self) -> SHFPPCSweeperConfig:
        return self._flatten(self._command_tree)

File: src/rust/sampled-event-handler/src/shfppc_sweeper_config_tracker.rs
// Copyright 2025 Zurich Instruments AG
// SPDX-License-Identifier: Apache-2.0

use std::rc::Rc;
use std::{cell::RefCell, ops::Deref};

use crate::shfppc_sweeper_config::SHFPPCSweeperConfig;
use codegenerator::ir::experiment::SweepCommand;

enum SweepNode {
    SweepCommand(SweepCommand),
    SweepStackFrame(Rc<RefCell<SweepStackFrame>>),
}

struct SweepStackFrame {
    count: u64,
    items: Vec<Rc<RefCell<SweepNode>>>,
}

enum SweepConfigNode {
    SweepCommand(SweepCommand),
    SHFPPCSweeperConfig(SHFPPCSweeperConfig),
}

pub(crate) struct SHFPPCSweeperConfigTracker {
    command_tree: Rc<RefCell<SweepStackFrame>>,
    stack: Vec<Rc<RefCell<SweepStackFrame>>>,
    finished: bool,
}

impl SHFPPCSweeperConfigTracker {
    pub fn new() -> Self {
        let command_tree = Rc::new(RefCell::new(SweepStackFrame {
            count: 1,
            items: Vec::new(),
        }));
        Self {
            command_tree: Rc::clone(&command_tree),
            stack: vec![command_tree],
            finished: false,
        }
    }

    fn current_frame(&self) -> &Rc<RefCell<SweepStackFrame>> {
        self.stack
            .last()
            .expect("Internal error: No stack frame available")
    }

    pub fn add_step(&mut self, sweep_command: SweepCommand) {
        self.current_frame()
            .borrow_mut()
            .items
            .push(Rc::new(RefCell::new(SweepNode::SweepCommand(
                sweep_command,
            ))));
    }

    pub fn enter_loop(&mut self, count: u64) {
        let new_frame = Rc::new(RefCell::new(SweepStackFrame {
            count,
            items: Vec::new(),
        }));
        let new_node = Rc::new(RefCell::new(SweepNode::SweepStackFrame(Rc::clone(
            &new_frame,
        ))));
        self.current_frame().borrow_mut().items.push(new_node);
        self.stack.push(new_frame);
    }

    pub fn exit_loop(&mut self) {
        let closed_frame = self.stack.pop().expect("Internal error: No frame to close");
        // If the frame we are closing is empty, we don't need to keep it
        if closed_frame.borrow().items.is_empty() {
            self.current_frame().borrow_mut().items.pop();
        }
    }

    /// Get the final SHFPPCSweeperConfig from the command tree.
    ///
    /// After calling this function, the command tree will be empty.
    /// This function may be called only once at the end of the configuration.
    ///
    pub fn finish(&mut self) -> Option<SHFPPCSweeperConfig> {
        assert!(
            !self.finished,
            "Internal error: finish() called multiple times"
        );
        self.finished = true;
        let (commands, count) = flatten(&self.command_tree);
        if commands.is_empty() {
            return None;
        }
        Some(SHFPPCSweeperConfig { count, commands })
    }
}

/// Remove any nesting from the command tree
///
/// After flattening, `items` will only hold `SweepCommand`s. The repetition
/// count may get adjusted to efficiently represent any internal loops.
///
fn flatten(command_tree: &Rc<RefCell<SweepStackFrame>>) -> (Vec<SweepCommand>, u64) {
    let command_tree = command_tree.borrow_mut();
    let mut count = command_tree.count;
    let mut commands = command_tree
        .items
        .iter()
        .map(|c| match c.borrow().deref() {
            SweepNode::SweepStackFrame(frame) => {
                assert!(
                    !frame.borrow().items.is_empty(),
                    "Internal error: Empty frames should have been dropped earlier"
                );
                let (commands, count) = flatten(frame);
                SweepConfigNode::SHFPPCSweeperConfig(SHFPPCSweeperConfig { count, commands })
            }
            SweepNode::SweepCommand(command) => SweepConfigNode::SweepCommand(command.clone()),
        })
        .collect::<Vec<_>>();

    let commands = match commands.as_mut_slice() {
            [SweepConfigNode::SHFPPCSweeperConfig(_)] => {
                if let SweepConfigNode::SHFPPCSweeperConfig(child_config) = commands.pop().unwrap() {
                    count *= child_config.count;
                    child_config.commands
                } else {
                    unreachable!("Internal error: Expected SHFPPCSweeperConfig")
                }
            }
            _ => commands
                .into_iter()
                .map(|node| match node {
                    SweepConfigNode::SweepCommand(command) => command,
                    _ => panic!("Internal error: Items must be SweepCommand. Nesting rolled loops is not supported."),
                })
                .collect(),
        };
    (commands, count)
}
